### PR TITLE
Replace aws sdk client

### LINKED ISSUE #16059: Replace `aws-sdk` with `@aws-sdk/client-s3` in `@directus/drive`
`@directus/drive` currently uses AWS SDK v2, which includes way more code than we need. For AWS SDK v3, AWS opted to ship individual client packages for the various services, meaning that we can vastly reduce the overall footprint of `@directus/drive` by switching to v3. That being said, every API call is done differently, so it's not just a drop-in replacement.

### PR DESCRIPTION
**Pull request by davidecarlier**
## Description

This update replaces the usage of aws-sdk in the drive-s3 package with @aws-sdk/client-s3. 

Fixes #16059 

## Type of Change

- [ ] Bugfix
- [x] Improvement
- [ ] New Feature
- [ ] Refactor / codestyle updates
- [ ] Other, please describe:

## Requirements Checklist

- [ ] New / updated tests are included
- [x] All tests are passing locally
- [x] Performed a self-review of the submitted code

If adding a new feature:

- [ ] Documentation was added/updated. PR:

### CODE REVIEW

File: packages/drive-s3/src/AmazonWebServicesS3Storage.ts [hunk 1]
```diff
@@ -147,14 +171,11 @@ export class AmazonWebServicesS3Storage extends Storage {
 		location = this._fullPath(location);
 
 		const params = { Key: location, Bucket: this.$bucket };
-
+		const result = await this.$driver.send(new GetObjectCommand(params));
+		const body = result.Body as NodeJS.ReadableStream;
```
Comment (paescuj) [inline packages/drive-s3/src/AmazonWebServicesS3Storage.ts:175]: This should probably go in to the `try...catch` block

Comment (davidecarlier) [reply packages/drive-s3/src/AmazonWebServicesS3Storage.ts]: Yep! fixed

File: packages/drive-s3/src/AmazonWebServicesS3Storage.ts [hunk 2]
```diff
@@ -223,13 +250,13 @@
 	public getUrl(location: string): string {
 		location = this._fullPath(location);
 
-		const { href } = this.$driver.endpoint;
+		const endpoint = this.$driver.config.endpointProvider({ Region: this.$region });
 
-		if (href.startsWith('https://s3.amazonaws')) {
+		if (endpoint.url.href.startsWith('https://s3.amazonaws')) {
```
[the remark this answers is not in the document]
Comment (davidecarlier) [reply packages/drive-s3/src/AmazonWebServicesS3Storage.ts]: I removed this if statement in my latest commit and fixed the use of endpointProvider to generate the URL correctly.

**Review by rijkvanzanten — CHANGES_REQUESTED** [review]
Seems like I'm getting an error when uploading files from Directus 🤔 

```
The \"path\" argument must be of type string. Received null

at AmazonWebServicesS3Storage._fullPath (/Users/rijkvanzanten/Developer/directus/directus/packages/drive-s3/dist/AmazonWebServicesS3Storage.js:53:61
```

**Comment by davidecarlier** [comment]
@rijkvanzanten can you please try again now?

**Comment by rijkvanzanten** [comment]
Different error now!

```
{
  "errors": [
    {
      "message": "Couldn't save file 9785ffee-5350-44ee-a6b1-2557e04809b5.jpg",
      "extensions": {
        "code": "SERVICE_UNAVAILABLE",
        "service": "files",
        "stack": "Error: Couldn't save file 9785ffee-5350-44ee-a6b1-2557e04809b5.jpg\n    at FilesService.uploadOne (/Users/rijkvanzanten/Developer/directus/directus/api/src/services/files.ts:77:10)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at Multipart.<anonymous> (/Users/rijkvanzanten/Developer/directus/directus/api/src/controllers/files.ts:90:23)"
      }
    }
  ]
}
```

from the API logs:

```
            Error code: TypeError
            Original stack:
            TypeError: Cannot read properties of null (reading 'length')
                at resolvedPath (/Users/rijkvanzanten/Developer/directus/directus/node_modules/.pnpm/@aws-sdk+smithy-client@3.201.0/node_modules/@aws-sdk/smithy-client/dist-cjs/resolve-path.js:8:24)
```

**Comment by davidecarlier** [comment]
I can't reproduce the bug 😕 
Can you please share your configurations?

**Comment by rijkvanzanten** [comment]
Using localhost minio as the S3 debugger (from the root folder of the repo `docker compose up -d minio`, with the following configuration:

```
STORAGE_LOCATIONS="minio"
STORAGE_MINIO_DRIVER="s3"
STORAGE_MINIO_KEY="minioadmin"
STORAGE_MINIO_SECRET="minioadmin"
STORAGE_MINIO_BUCKET="directus"
STORAGE_MINIO_ENDPOINT="localhost:5106"
```

Make sure to login to minio on localhost:5112 with minioadmin/minioadmin to create the `directus` bucket first

**Comment by davidecarlier** [comment]
Hi @rijkvanzanten, I fixed a few things in the configuration of custom endpoints.
_On my machine_, it works both with a real s3 account and with minio


Side note, I tried to use minio on another directus instance (9.20.4) and it doesn't work:
`Inaccessible host: `directus.localhost' at port `undefined'`

So, I added directus.localhost to my hosts but now I get the error:
`MalformedXML: The XML you provided was not well-formed or did not validate against our published schema.`
Am I missing something about minio?

**Pull request closed without merging**